File: app/src/main/java/net/opendasharchive/openarchive/fragments/MediaGridFragment.java
package net.opendasharchive.openarchive.fragments;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import net.opendasharchive.openarchive.R;
import net.opendasharchive.openarchive.db.Collection;
import net.opendasharchive.openarchive.db.Media;
import net.opendasharchive.openarchive.db.MediaAdapter;
import net.opendasharchive.openarchive.media.PreviewMediaListActivity;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import kotlinx.coroutines.CoroutineScope;
import kotlinx.coroutines.Dispatchers;

public class MediaGridFragment extends MediaListFragment {

    private int numberOfColumns = 4;
    private HashMap<Long, MediaAdapter> mAdapters;
    private HashMap<Long,SectionViewHolder> mSection;

    private View mMainView;

    private View mMediaHint;

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public MediaGridFragment() {
        super();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        mAdapters = new HashMap<>();
        mSection = new HashMap<>();

        mMainView = inflater.inflate(R.layout.fragment_media_list, container, false);
        mMainView.setTag(TAG);

        LinearLayout mainContainer = mMainView.findViewById(R.id.mediacontainer);

        mMediaHint = mMainView.findViewById(R.id.add_media_hint);

        List<Collection> listCollections = Collection.Companion.getAllAsList();

        boolean addedView = false;
        for (Collection coll : listCollections) {

            List<Media> listMedia = Media.Companion.getMediaByProjectAndCollection(mProjectId, coll.getId());
            if (listMedia.size() > 0)
            {
                if (!addedView)
                {
                    mainContainer.removeAllViews();
                    addedView = true;
                }

                View view = createMediaList(mainContainer, coll, listMedia);
                mainContainer.addView(view);

            }
        }

        if (mMediaHint != null)
            mMediaHint.setVisibility(addedView? View.GONE : View.VISIBLE);

        return mMainView;
    }

    private class SectionViewHolder {
        View mediaSection;
        RecyclerView.Recycler mediaGrid;
        TextView sectionStatus;
        TextView sectionTimestamp;
        View action;
    }

    private View createMediaList (LinearLayout mainContainer, Collection coll, List<Media> listMedia)
    {
        GridLayoutManager lMan = new GridLayoutManager(getActivity(), numberOfColumns) {
            @Override
            public boolean canScrollVertically() {
                return false;
            }
        };

        SectionViewHolder holder = new SectionViewHolder();

        holder.mediaSection = getLayoutInflater().inflate(R.layout.fragment_media_list_section, null);
        RecyclerView rView = holder.mediaSection.findViewById(R.id.recyclerview);
        rView.setHasFixedSize(true);
        rView.setLayoutManager(lMan);

        holder.sectionStatus = holder.mediaSection.findViewById(R.id.sectionstatus);
        holder.sectionTimestamp = holder.mediaSection.findViewById(R.id.sectiontimestamp);
        holder.action = holder.mediaSection.findViewById(R.id.action_next);

        setSectionHeaders(coll, listMedia, holder);

        ArrayList<Media> listMediaArray = new ArrayList<>(listMedia);
        MediaAdapter mediaAdapter = new MediaAdapter(getActivity(), R.layout.activity_media_list_square, listMediaArray, rView, viewHolder -> {

        });
        rView.setAdapter(mediaAdapter);
        mAdapters.put(coll.getId(),mediaAdapter);
        mSection.put(coll.getId(),holder);

        return holder.mediaSection;
    }

    public void updateItem (long mediaId, long progress)
    {
        for (MediaAdapter adapter : mAdapters.values())
            adapter.updateItem(mediaId, progress);

    }

    public void refresh ()
    {

        LinearLayout mainContainer = mMainView.findViewById(R.id.mediacontainer);

        List<Collection> listCollections = Collection.Companion.getAllAsList();

        for (Collection coll : listCollections) {

            List<Media> listMedia = Media.Companion.getMediaByProjectAndCollection(mProjectId, coll.getId());

            MediaAdapter adapter = mAdapters.get(coll.getId());
            SectionViewHolder holder = mSection.get(coll.getId());

            ArrayList listMediaArray = new ArrayList(listMedia);

            if (adapter != null)
            {
                adapter.updateData(listMediaArray);

                setSectionHeaders(coll, listMedia, holder);

            }
            else if (listMedia.size() > 0)
            {
                View view = createMediaList(mainContainer, coll,listMedia);
                mainContainer.addView(view,0);

                if (mMediaHint != null)
                    mMediaHint.setVisibility(View.GONE);
            }
        }

    }

    private void setSectionHeaders (Collection coll, List<Media> listMedia, SectionViewHolder holder)
    {
        for (Media media : listMedia)
        {
            if (media.getStatus() == Media.STATUS_LOCAL)
            {
                holder.sectionStatus.setText(R.string.status_ready_to_upload);
                holder.sectionTimestamp.setText(listMedia.size() + getString(R.string.label_items));
                holder.action.setVisibility(View.VISIBLE);
                holder.action.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        startActivity(new Intent(getActivity(), PreviewMediaListActivity.class));
                    }
                });
                break;

            }
            else if (media.getStatus() == Media.STATUS_QUEUED || media.getStatus() == Media.STATUS_UPLOADING)
            {
                holder.sectionStatus.setText(R.string.header_uploading);

                int uploadedCount = 0;
                for (Media localMedia : listMedia)
                    if (localMedia.getStatus() == Media.STATUS_UPLOADED)
                        uploadedCount++;

                holder.sectionTimestamp.setText(uploadedCount + " " + getString(R.string.label_out_of) + " " + listMedia.size() + ' ' + getString(R.string.label_items_uploaded));


                holder.action.setVisibility(View.GONE);
                break;
            }
            else
            {
                holder.sectionStatus.setText(listMedia.size() + " " + getString(R.string.label_items_uploaded));
                if (coll.getUploadDate() != null)
                    holder.sectionTimestamp.setText(coll.getUploadDate().toLocaleString());
                else if (listMedia.size() > 0 && listMedia.get(0).getUploadDate() != null)
                    holder.sectionTimestamp.setText(listMedia.get(0).getUploadDate().toString());

                holder.action.setVisibility(View.GONE);
            }
        }
    }
}

File: app/src/main/java/net/opendasharchive/openarchive/features/media/grid/MediaGridFragment.kt
package net.opendasharchive.openarchive.features.media.grid

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.children
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson
import net.opendasharchive.openarchive.R
import net.opendasharchive.openarchive.databinding.FragmentMediaListBinding
import net.opendasharchive.openarchive.databinding.FragmentMediaListSectionBinding
import net.opendasharchive.openarchive.db.*
import net.opendasharchive.openarchive.db.Collection
import net.opendasharchive.openarchive.features.media.SectionViewHolder
import net.opendasharchive.openarchive.features.media.list.MediaListFragment
import net.opendasharchive.openarchive.features.media.preview.PreviewMediaListActivity
import net.opendasharchive.openarchive.features.media.preview.PreviewMediaListViewModel
import net.opendasharchive.openarchive.features.media.preview.PreviewMediaListViewModelFactory
import net.opendasharchive.openarchive.util.Prefs
import net.opendasharchive.openarchive.util.extensions.cloak
import net.opendasharchive.openarchive.util.extensions.toggle
import java.text.DateFormat
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.List
import kotlin.collections.filter
import kotlin.collections.firstOrNull
import kotlin.collections.forEach
import kotlin.collections.isNotEmpty
import kotlin.collections.listOf
import kotlin.collections.set

class MediaGridFragment : MediaListFragment() {

    private val numberOfColumns = 4
    private var mAdapters = HashMap<Long, MediaAdapter>()
    private var mSection = HashMap<Long, SectionViewHolder>()

    private lateinit var mBinding: FragmentMediaListBinding
    private lateinit var viewModel: MediaGridViewModel
    private lateinit var previewViewModel: PreviewMediaListViewModel

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        mBinding = FragmentMediaListBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(this)[MediaGridViewModel::class.java]

        val context = requireNotNull(activity?.application)
        val viewModelFactory = PreviewMediaListViewModelFactory(context)
        previewViewModel = ViewModelProvider(this, viewModelFactory)[PreviewMediaListViewModel::class.java]
        previewViewModel.observeValuesForWorkState(activity as AppCompatActivity)

        observeData()
        viewModel.getAllCollection()
        return mBinding.root
    }

    private fun observeData() {
        viewModel.collections.observe(viewLifecycleOwner) {
            initLayout(it)
        }
    }

    private fun initLayout(collections: List<Collection>?) {
        mAdapters = HashMap()
        mSection = HashMap()

        mBinding.root.tag = TAG

        var addedView = false

        for (collection in collections ?: emptyList()) {
            if (collection.projectId != getProjectId()) continue

            val media = collection.media

            if (media.isEmpty()) continue

            if (!addedView) {
                for (view in mBinding.mediacontainer.children) {
                    if (view != mBinding.addMediaHint) {
                        mBinding.mediacontainer.removeView(view)
                    }
                }

                addedView = true
            }

            mBinding.mediacontainer.addView(createMediaList(collection, media))
        }

        mBinding.addMediaHint.toggle(mBinding.mediacontainer.childCount < 2)
    }

    private fun performBatchUpload(listMedia: List<Media>) {
        for (media in listMedia) {
            media.sStatus = Media.Status.Queued
            media.save()
        }
        previewViewModel.applyMedia()
    }

    private fun createMediaList(collection: Collection, listMedia: List<Media>): View {
        val holder = SectionViewHolder(FragmentMediaListSectionBinding.inflate(layoutInflater))

        holder.apply {
            mediaSection.recyclerview.setHasFixedSize(true)
            mediaSection.recyclerview.layoutManager = GridLayoutManager(activity, numberOfColumns)

            setSectionHeaders(collection, listMedia, this)

            val listMediaArray = ArrayList(listMedia)
            val mediaAdapter = MediaAdapter(
                requireContext(),
                R.layout.activity_media_list_square,
                listMediaArray,
                mediaSection.recyclerview,
                object : OnStartDragListener {
                    override fun onStartDrag(viewHolder: RecyclerView.ViewHolder?) {

                    }
                }, onDelete = {
                    refresh()
                }, onUpload = {
                    if (Space.current?.tType == Space.Type.WEBDAV) {
                        // TODO: WTF is this DUPLICATED special casing here?!?
                        if (Space.current?.host?.contains("https://sam.nl.tab.digital") == true) {
                            val availableSpace = getAvailableStorageSpace(it)
                            val totalUploadsContent = availableSpace.first
                            val totalStorage = availableSpace.second
                            if(totalStorage < totalUploadsContent) {
                                Toast.makeText(activity, getString(R.string.upload_files_error), Toast.LENGTH_SHORT).show()
                            }
                            else {
                                performBatchUpload(it)
                            }
                        }
                        else {
                            //for NON nextcloud providers
                            performBatchUpload(it)
                        }
                    }
                    else {
                        //for non webDAV protocols.
                        performBatchUpload(it)
                    }
                })

            mediaSection.recyclerview.adapter = mediaAdapter
            mAdapters[collection.id] = mediaAdapter
            mSection[collection.id] = this
        }

        return holder.mediaSection.root
    }

    private fun getAvailableStorageSpace(listMedia: List<Media>): Pair<Double, Long> {
        val nextCloudModel = Gson().fromJson(Prefs.nextCloudModel, WebDAVModel::class.java)
        var totalUploadsContent = 0.0
        for (media in listMedia) {
            totalUploadsContent += media.contentLength
        }

        val totalStorage = nextCloudModel.ocs.data.quota.total - nextCloudModel.ocs.data.quota.used
        return Pair(totalUploadsContent, totalStorage)
    }

    override fun updateItem(mediaId: Long, progress: Long) {
        for (adapter in mAdapters.values) adapter.updateItem(mediaId, progress)
    }

    override fun refresh() {
        Collection.getAll().forEach { collection ->
            val media = if (collection.projectId == getProjectId()) collection.media else listOf()

            val adapter = mAdapters[collection.id]
            val holder = mSection[collection.id]

            if (adapter != null) {
                adapter.updateData(ArrayList(media))
                if (holder != null) setSectionHeaders(collection, media, holder)
            }
            else if (media.isNotEmpty()) {
                val view = createMediaList(collection, media)

                mBinding.mediacontainer.addView(view, 0)
            }
        }

        mBinding.addMediaHint.toggle(mBinding.mediacontainer.childCount < 2)
    }

    @SuppressLint("SetTextI18n")
    private fun setSectionHeaders(
        collection: Collection,
        listMedia: List<Media>?,
        holder: SectionViewHolder
    ) {
        holder.sectionStatus.text = ""
        holder.sectionTimestamp.text = ""

        val df = DateFormat.getDateTimeInstance()

        listMedia?.forEach { media ->
            when (media.sStatus) {
                Media.Status.Local -> {
                    holder.sectionStatus.text = getString(R.string.status_ready_to_upload)
                    holder.sectionTimestamp.text = "${listMedia.size} ${getString(R.string.label_items)}"
                    holder.action.cloak()
                    holder.action.setOnClickListener {
                        startActivity(Intent(requireActivity(), PreviewMediaListActivity::class.java))
                    }

                    return@forEach
                }
                Media.Status.Queued, Media.Status.Uploading -> {
                    holder.sectionStatus.text = getString(R.string.header_uploading)

                    val count = listMedia.filter { it.sStatus == Media.Status.Uploaded }.size

                    holder.sectionTimestamp.text = getString(R.string.__out_of___items_uploaded, count, listMedia.size)
                    holder.action.cloak()

                    return@forEach
                }
                Media.Status.Uploaded -> {
                    val count = listMedia.filter { it.sStatus == Media.Status.Uploaded }.size

                    if (count == listMedia.size) {
                        holder.sectionStatus.text = getString(R.string.__items_uploaded, listMedia.size)
                        holder.action.cloak()
                    }
                    else {
                        holder.sectionStatus.text = getString(R.string.__out_of___items_uploaded, count, listMedia.size)
                        holder.action.cloak()
                    }

                    collection.uploadDate?.let { holder.sectionTimestamp.text = df.format(it) }
                }
                Media.Status.Error -> {
                    val count = listMedia.filter { it.sStatus == Media.Status.Error }.size

                    if (count == listMedia.size) {
                        holder.sectionStatus.text = getString(R.string.__items_failed_to_upload, listMedia.size)
                    }
                    else {
                        holder.sectionStatus.text = getString(R.string.__out_of___items_uploaded, count, listMedia.size)
                    }

                    collection.uploadDate?.let { holder.sectionTimestamp.text = df.format(it) }

                    holder.action.cloak()
                }
                else -> {
                    holder.sectionStatus.text = getString(R.string.__items_uploaded, listMedia.size)

                    collection.uploadDate?.let { holder.sectionTimestamp.text = df.format(it) }
                        ?: run {
                            listMedia.firstOrNull()?.uploadDate?.let {
                                holder.sectionTimestamp.text = df.format(it)
                            }
                        }

                    holder.action.cloak()
                }
            }
        }
    }
}
